cpp/src/arrow/dataset/partition.cc

```diff
@@ -82,6 +82,14 @@ std::shared_ptr<Partitioning> Partitioning::Default() {
 
     std::string type_name() const override { return "default"; }
 
+    bool Equals(const Partitioning& other, bool check_metadata) const override {
+      if (this == &other) {
+        return true;
+      }
+      return checked_cast<const DefaultPartitioning&>(other).type_name() == type_name() &&
```

Comment:
I agree.  If you received a pointer type `Partitioning*` you could do a `dynamic_cast` instead of the `type_name` comparison.  However, a failed `dynamic_cast` is an exception when you have a reference type and it's probably better to just compare `type_name`.

The comparison to `type_name` is basically doing the same thing as a cast check anyways.

The reason a `checked_cast` is a bad idea is that it should be ok for a user to do:

```
DefaultPartitioning one;
KeyValuePartitioning two(schema);
if (one.Equals(two)) {
  std::cout << "equal" << std::endl;
}
```

This `Equals` method should always return false so this toy example is somewhat meaningless.  However, it could be useful if the actual partitions were based on file metadata or something dynamically calculated.  If there is a `checked_cast` then, instead of false, we will get an abort in debug mode and potentially a segmentation fault in release mode.

Reply:
Yes your absolutely right, sorry I missed this. Thanks for pointing this out.